docs/proposals/20210222-kubelet-authentication.md

```diff
@@ -0,0 +1,376 @@
+
+Cluster API is responsible for partially verifying node identity with the following conditions:
+
+- A corresponding machine object exist for the CSR's `.spec.Username` (`system:nodes:<nodename>`) (providing the value is deferred to infrastructure provider)
+- The Machine must have conditions BootstrapReady.
+- The Kubernetes CSR spec has the needed groups
+- The Kubernetes CSR spec is limited to needed usages (e.g. client auth)
+- The Kubernetes CSR spec is limited to needed extensions (e.g. no CA extension)
+- Parse the CSR and verify that the CN is the same as .spec.username
+- Parse the CSR and verify that the Organization is the same as .spec.Groups
+- Parse the CSR and ensure that no SANs are appended for kubelet client certificates
+
+#### CSR format used by kubelet-authenticator
+We propose the introduction of X.509 extension attributes based
+on those reserved for the Kubernetes GCP cloud provider within Google’s organization ID allocation.
+
+We will request via SIG Architecture or CNCF to apply for an [IANA OID registration block][iana-issue] for the Kubernetes project.
+
+| OID Suffix | Name                                | Description                                                                                                                                                                                                                            |
+| ---------- | ----------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| 2.1.21     | KubernetesNodeProviderIdentifierOID | An identifier for the machine, should be the same or a derivative of the node provider ID. This is the equivalent of Google’s CloudComputeInstanceIdentifierOID, which we can reuse for a proof of concept (1.3.6.1.4.1.11129.2.1.21). |
+
+#### CSR PEM Blocks
+
+The following blocks will be added to CSRs following Section 2 of [RFC7468].
+
+| Block Name                                 | Description                                          |
+| ------------------------------------------ | ---------------------------------------------------- |
+| KUBELET AUTHENTICATOR ATTESTATION PROVIDER | string describing the attestation provider           |
+| KUBELET AUTHENTICATOR ATTESTATION DATA     | the actual attestation data to perform validation on |
+
+#### Attestation data
+
+Attestation data will be appended with the following headers and footers and MUST
+be base64 encoded.
+
+Example CSR:
+```
+-----BEGIN ATTESTATION DATA-----
+S25vd2luZyBtZSBBbGFuIFBhcnRyaWRnZSwga25vd2luZyB5b3UgS3ViZXJuZXRlcyBjbHVzdGVyLCBhaGEh
+-----END ATTESTATION DATA-----
+```
+The format of the attestation block is left to the provider. The Appendix section provides examples.
```

Comment:
I think the Appendix section is missing

Reply:
Going to defer to infra provider proposals for those. Was in the google doc version.